### PR TITLE
Unique goods receipt

### PR DESCRIPTION
**Pull request by thenosewizard**

### CODE REVIEW

File: src/main/java/seedu/address/MainApp.java [hunk 1]
```diff
@@ -96,20 +92,20 @@ private Model initModelManager(Storage storage, ReadOnlyUserPrefs userPrefs) {
...
-            initialGoodsData = goodsReceiptList.orElseGet(ArrayList::new);
+            //initialGoodsData = goodsReceiptList.orElseGet()
         } catch (DataLoadingException e) {
             logger.warning("Creating a new data file " + storage.getGoodsFilePath()
                     + " populated with no Goods.");
-            initialGoodsData = new ArrayList<>();
+            initialGoodsData = new ReceiptLog();
         }
 
         return new ModelManager(initialData, userPrefs, initialGoodsData);
```
Comment (thenosewizard) [inline src/main/java/seedu/address/MainApp.java:111]: @stephaneleong Do check if the logic here is feasible, where sample data is not ready yet

File: src/main/java/seedu/address/storage/CsvGoodsStorage.java [hunk 2]
```diff
@@ -33,27 +36,30 @@ public Path getGoodsFilePath() {
...
 
     @Override
-    public Optional<List<GoodsReceipt>> readGoods(Path filePath) throws DataLoadingException {
+    public Optional<ReadOnlyReceiptLog> readGoods(Path filePath) throws DataLoadingException {
         requireNonNull(filePath);
-        return CsvUtil.readCsvFile(filePath, GoodsReceipt.class);
+
+        ReceiptLog r = new ReceiptLog();
+        r.setReceipts(CsvUtil.readCsvFile(filePath, GoodsReceipt.class).get());
+        return Optional.of(r);
```
Comment (thenosewizard) [inline src/main/java/seedu/address/storage/CsvGoodsStorage.java:49]: @stephaneleong logic update

File: src/test/java/seedu/address/storage/CsvGoodsStorageTest.java [hunk 3]
```diff
@@ -53,9 +55,8 @@ public void readGoods_nullFilePath_throwsNullPointerException() {
     }
 
     @Test
-    public void readGoods_missingFile_emptyList() throws IOException, DataLoadingException {
-        Optional<List<GoodsReceipt>> goods = csvGoodsStorage.readGoods(Paths.get("NonExistentFile.csv"));
-        assertTrue(goods.isEmpty());
+    public void readGoods_missingFile_emptyList() throws IOException, DataLoadingException, NoSuchElementException {
+        assertThrows(NoSuchElementException.class, () -> csvGoodsStorage.readGoods(Paths.get("NonExistentFile.csv")));
     }
```
Comment (thenosewizard) [inline src/test/java/seedu/address/storage/CsvGoodsStorageTest.java:60]: @stephaneleong test update

File: src/main/java/seedu/address/MainApp.java [hunk 4]
```diff
@@ -96,20 +95,20 @@ private Model initModelManager(Storage storage, ReadOnlyUserPrefs userPrefs) {
...
-            initialGoodsData = goodsReceiptList.orElseGet(ArrayList::new);
+            initialGoodsData = goodsReceiptList.orElseGet(ReceiptLog::new);
         } catch (DataLoadingException e) {
             logger.warning("Creating a new data file " + storage.getGoodsFilePath()
                     + " populated with no Goods.");
-            initialGoodsData = new ArrayList<>();
+            initialGoodsData = new ReceiptLog();
         }
 
         return new ModelManager(initialData, userPrefs, initialGoodsData);
```
Comment (thenosewizard) [inline src/main/java/seedu/address/MainApp.java:114]: Updated logic for from List<GoodsReceipt> to ReadOnlyReceiptLog

File: src/main/java/seedu/address/storage/CsvGoodsStorage.java [hunk 5]
```diff
@@ -33,27 +34,30 @@ public Path getGoodsFilePath() {
...
         return readGoods(filePath);
     }
 
     @Override
-    public Optional<List<GoodsReceipt>> readGoods(Path filePath) throws DataLoadingException {
+    public Optional<ReadOnlyReceiptLog> readGoods(Path filePath) throws DataLoadingException {
         requireNonNull(filePath);
-        return CsvUtil.readCsvFile(filePath, GoodsReceipt.class);
+
+        ReceiptLog r = new ReceiptLog();
```
Comment (stephaneleong) [inline src/main/java/seedu/address/storage/CsvGoodsStorage.java:45]: Logic is correct, better not to use scratch variable r for ReceptLog. giving it a more meaningful name is more preferred

File: src/main/java/seedu/address/storage/CsvGoodsStorage.java (hunk 2 above)
Comment (stephaneleong) [reply src/main/java/seedu/address/storage/CsvGoodsStorage.java:47]: logic looks good

File: src/test/java/seedu/address/storage/CsvGoodsStorageTest.java (hunk 3 above)
Comment (stephaneleong) [reply src/test/java/seedu/address/storage/CsvGoodsStorageTest.java:55]: okay, test case name need to updated readGoods_missingFile_exceptionThrown()

File: src/test/java/seedu/address/testutil/TypicalGoods.java [hunk 6]
```diff
@@ -63,8 +65,11 @@ public static List<Goods> getTypicalGoods() {
     }
 
 
-    public static List<GoodsReceipt> getTypicalGoodsReceipts() {
-        return new ArrayList<>(Arrays.asList(
+    public static ReadOnlyReceiptLog getTypicalGoodsReceipts() {
+
+        ReceiptLog r = new ReceiptLog();
```
Comment (stephaneleong) [inline src/test/java/seedu/address/testutil/TypicalGoods.java:70]: same as the comment before preference on not using scratch variables

**Review by stephaneleong — COMMENTED** [review]
Good Job on adding the new abstraction to ensure goods are not duplicated. needs some clean up on java coding standards i.e. minor renaming of scratch variables and test case name

File: src/main/java/seedu/address/model/ReceiptLog.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+     * Adds a receipt to the receipt log.
+     * The receipt must not already exist in the receipt log
+     */
+    public void addReceipt(GoodsReceipt add) {
+        receipts.add(add);
+    }
+
+
+    /**
+     * deletes a receipt to the receipt log.
```
Comment (jialegitgud) [inline src/main/java/seedu/address/model/ReceiptLog.java:66]: Capitalization error here

File: src/main/java/seedu/address/model/ReceiptLog.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        this.receipts.setReceipts(receipts);
+    }
+
+    /**
+     * Finds a receipt
+     */
+    public List<GoodsReceipt> findReceipts(Predicate<GoodsReceipt> predicate) {
+        return StreamSupport
+                .stream(receipts.spliterator(), false)
+                .filter(predicate).collect(Collectors.toList());
```
Comment (jialegitgud) [inline src/main/java/seedu/address/model/ReceiptLog.java:85]: Would be easier to use .toList() directly, so we can remove one import. Also provides additional safety as streams.toList() returns an immutable list.

File: src/main/java/seedu/address/model/goodsReceipt/exceptions/DuplicateReceiptException.java [hunk 9]
```diff
@@ -0,0 +1,10 @@
+package seedu.address.model.goodsReceipt.exceptions;
+
+/**
+ * Signals that the operation will result in duplicate goodsReceipts
+ */
+public class DuplicateReceiptException extends RuntimeException {
```
Comment (jialegitgud) [inline src/main/java/seedu/address/model/goodsReceipt/exceptions/DuplicateReceiptException.java:6]: Should this be a checked exception instead?

File: src/main/java/seedu/address/model/goodsReceipt/UniqueGoodsReceiptList.java [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+     * Returns true if the list contains an equivalent receipt as the given argument.
+     */
+    public boolean contains(GoodsReceipt toCheck) {
+        requireNonNull(toCheck);
+        return internalList.stream().anyMatch(x -> x.isSameReceipt(toCheck));
+    }
+
+    /**
+     * Adds a person to the list.
+     * The person must not already exist in the list.
```
Comment (jialegitgud) [inline src/main/java/seedu/address/model/goodsReceipt/UniqueGoodsReceiptList.java:39]: Inaccurate comments here

**Review by jialegitgud — COMMENTED** [review]
As with Stephane's comments, has some minor code quality errors but overall looks good.

**Pull request merged**